From 45d5f35a8333062ad32605e958037ef48b413c3a Mon Sep 17 00:00:00 2001
From: Denys Vlasenko <dvlasenk@redhat.com>
Date: Wed, 26 Mar 2014 18:05:47 -0400
Subject: [tty] Set correct tty name in 'active' sysfs attribute

Message-id: <1395857147-1290-1-git-send-email-dvlasenk@redhat.com>
Patchwork-id: 78616
O-Subject: [PATCH RHEL7 BZ 1066403] tty: Set correct tty name in 'active' sysfs attribute
Bugzilla: 1066403
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Hendrik Brueckner <brueckner@redhat.com>
RH-Acked-by: Aristeu Rozanski <aris@redhat.com>

BZ:
https://bugzilla.redhat.com/bugzilla/show_bug.cgi?id=1066403

Brew build:
https://brewweb.devel.redhat.com/taskinfo?taskID=7258214

Upstream status of the patch:
upstream tty-next commit 723abd87f6e536f1353c8f64f621520bc29523a3
Author: Hannes Reinecke <hare@suse.de>
Date:   Thu Feb 27 12:30:51 2014 +0100

    tty: Set correct tty name in 'active' sysfs attribute

    The 'active' sysfs attribute should refer to the currently active tty
    devices the console is running on, not the currently active console. The
    console structure doesn't refer to any device in sysfs, only the tty the
    console is running on has. So we need to print out the tty names in
    'active', not the console names.

    There is one special-case, which is tty0. If the console is directed to
    it, we want 'tty0' to show up in the file, so user-space knows that the
    messages get forwarded to the active VT. The ->device() callback would
    resolve tty0, though. Hence, treat it special and don't call into the VT
    layer to resolve it (plymouth is known to depend on it).

Test Status:
Only compile-tested.

Signed-off-by: Denys Vlasenko <dvlasenk@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/tty/tty_io.c b/drivers/tty/tty_io.c
index 59d26ef..3723c0e 100644
--- a/drivers/tty/tty_io.c
+++ b/drivers/tty/tty_io.c
@@ -1267,12 +1267,13 @@ static void pty_line_name(struct tty_driver *driver, int index, char *p)
  *
  * Locking: None
  */
-static void tty_line_name(struct tty_driver *driver, int index, char *p)
+static ssize_t tty_line_name(struct tty_driver *driver, int index, char *p)
 {
  if (driver->flags & TTY_DRIVER_UNNUMBERED_NODE)
-  strcpy(p, driver->name);
+  return sprintf(p, "%s", driver->name);
  else
-  sprintf(p, "%s%d", driver->name, index + driver->name_base);
+  return sprintf(p, "%s%d", driver->name,
+          index + driver->name_base);
 }
 
 /**
@@ -3538,9 +3539,19 @@ static ssize_t show_cons_active(struct device *dev,
   if (i >= ARRAY_SIZE(cs))
    break;
  }
- while (i--)
-  count += sprintf(buf + count, "%s%d%c",
-     cs[i]->name, cs[i]->index, i ? ' ':'\n');
+ while (i--) {
+  int index = cs[i]->index;
+  struct tty_driver *drv = cs[i]->device(cs[i], &index);
+
+  /* don't resolve tty0 as some programs depend on it */
+  if (drv && (cs[i]->index > 0 || drv->major != TTY_MAJOR))
+   count += tty_line_name(drv, index, buf + count);
+  else
+   count += sprintf(buf + count, "%s%d",
+      cs[i]->name, cs[i]->index);
+
+  count += sprintf(buf + count, "%c", i ? ' ':'\n');
+ }
  console_unlock();
 
  return count;
-- 
1.7.1